### PR TITLE
feat(cli/fmt_errors): Truncate error location when source is "data:application/javascript..." Data URL

### PR DESCRIPTION
**Pull request by mikehw**
Addresses #9501. 

When the source of an Error is within a [Data URL](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs), truncate the "file name" in the formatted error location to be something like `data:application/javascript;base64,Y29uc...pOw==`.

Adds tests to cover cli and `core/error.js` formatting.

### CODE REVIEW

File: cli/tests/error_028_data_error_format_js.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+(async () => {
+  await import(
+    "data:application/javascript;base64,Y29uc29sZS5lcnJvcihFcnJvcigiYmFkIikpOw=="
```
Comment (nayeemrmn) [inline cli/tests/error_028_data_error_format_js.ts:3]: I think we should be a lot more generous as to how many characters are allowed before truncation happens. How about >120?

File: cli/fmt_errors.rs [hunk 2]
```diff
@@ -22,7 +22,19 @@ pub fn format_location(frame: &JsStackFrame) -> String {
...
+        let data_length = data_pieces.1.len();
+        if let Some(data_start) = data_pieces.1.get(0..5) {
+          if let Some(data_end) = data_pieces.1.get(data_length - 5..) {
+            fmt_file_name =
+              format!("{},{}...{}", data_pieces.0, data_start, data_end)
+          }
+        }
+      }
+    }
+    result += &cyan(fmt_file_name).to_string();
```
Comment (lucacasonato) [inline cli/fmt_errors.rs:37]: This is very brittle - please use the data URL parser for this. Not all data URLs are contain base64 data. They can also contain plain text.

File: cli/tests/error_028_data_error_format_js.ts (hunk 1 above)
Comment (mikehw) [reply cli/tests/error_028_data_error_format_js.ts:3]: Added a threshold of 150 to match `SOURCE_ABBREV_THRESHOLD`

File: cli/fmt_errors.rs (hunk 2 above)
Comment (mikehw) [reply cli/fmt_errors.rs]: Updated Rust and Javascript versions to use URL parsers. Modified the tests to demonstrate that it works the same regardless of data encoding.

**Comment by bartlomieju** [comment]
@mikehw sorry for the delay, I'll try to land this PR this week.

File: cli/fmt_errors.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/fmt_errors.rs]: I agree with Luca that this bit should use data URL parser. We already have dependency on `data_url` crate, here's an example how it's used in other parts of the codebase: https://github.com/denoland/deno/blob/e02403bb4590da6823eb956bcbc89e9a937d431d/cli/file_fetcher.rs#L330-L336

**Comment by bartlomieju** [comment]
I'm not very happy that this PR requires changes to CLI/runtime and core. @nayeemrmn do you have any ideas how we could handle it more gracefully?

**Comment by nayeemrmn** [comment]
@bartlomieju I reimplemented this in #12127 as a cli-core hook same as source mapping.

**Comment by bartlomieju** [comment]
Superseded by https://github.com/denoland/deno/pull/12127

**Pull request closed without merging**